src/db/pool.cr

```diff
@@ -34,24 +38,77 @@ module DB
     @availability_channel : Channel(Nil)
     # signal how many existing connections are waited for
     @waiting_resource : Int32
+    # communicate when the pool is closed, to kill the Connection Reaper
+    @close_channel : Channel(Nil)
     # global pool mutex
     @mutex : Mutex
 
-    def initialize(@initial_pool_size = 1, @max_pool_size = 0, @max_idle_pool_size = 1, @checkout_timeout = 5.0,
-                   @retry_attempts = 1, @retry_delay = 0.2, &@factory : -> T)
+    def initialize(
+      @initial_pool_size = 1,
+      @max_pool_size = 0,
+      @max_idle_pool_size = 1,
+      @checkout_timeout = 5.0,
+      @retry_attempts = 1,
+      @retry_delay = 0.2,
+      @reaping_delay = 60.0,
+      @reaping_frequency = 60.0,
+      &@factory : -> T
+    )
       @availability_channel = Channel(Nil).new
+      @close_channel = Channel(Nil).new
       @waiting_resource = 0
       @inflight = 0
       @mutex = Mutex.new
 
       @initial_pool_size.times { build_resource }
     end
 
+    def reap_connection!(connection : T)
+      sync do
+        return unless @idle.includes? connection
+
+        delete connection
+      end
+
+      connection.check
+
+      sync do
+        @total << connection
+        @idle << connection
+      end
+    rescue exc : PoolResourceLost(T)
+      connection.close
+      sync { add_resource! if can_increase_pool? }
+    end
+
+    def add_resource! : T
+      @inflight += 1
+      unsync { build_resource }
+    ensure # prevent inflight from incrementing if build_resource fails
+      @inflight -= 1
+    end
+
+    def start_reaper!
+      spawn(name: "connection_reaper") do
+        sleep @reaping_delay
+
+        loop do
+          select
+          when @close_channel.receive?
+            break
+          when timeout @reaping_frequency.seconds
+            @idle.each { |connection| reap_connection! connection }
```

Comment:
revisit to see if a delay method is worthwhile adding back w/the current implementation